test/common/TestCommonEqEpsilon.hs

```diff
@@ -34,6 +37,134 @@ eqEpsilonRef = unsafePerformIO $ newIORef eqEpsilonDefault
 setEpsilonEq :: EqEpsilon -> IO ()
 setEpsilonEq (EqEpsilon x) = atomicWriteIORef eqEpsilonRef x
 
+----------------------------------------------------------------------------
+-- Helper functions
+----------------------------------------------------------------------------
+
+go_assert_list :: forall a. ()
+               => (a -> a -> Assertion) -- ^ The function used to make an assertion on two elements (expected, actual)
+               -> [a]                   -- ^ The expected value
+               -> [a]                   -- ^ The actual value
+               -> Assertion
+go_assert_list _ [] [] = assertBool "" True
+go_assert_list _ [] (_:_) = assertFailure "More list elements than expected!"
+go_assert_list _ (_:_) [] = assertFailure "Less list elements than expected!"
+go_assert_list mk (head_exp:tail_exp) (head_act:tail_act) =
+      mk head_exp head_act >> go_assert_list mk tail_exp tail_act
+
+assert_list :: forall a. ()
+            => (a -> a -> Assertion) -- ^ The function used to make an assertion on two elements (expected, actual)
+            -> [a]                   -- ^ The expected value
+            -> [a]                   -- ^ The actual value
+            -> Assertion
+assert_list make_assert expected actual =
+  if lenE == lenA then
+    go_assert_list make_assert expected actual
+  else
+    assertFailure $ "List too " ++ (if lenE < lenA then "long" else "short") ++ ": expected " ++ show lenE ++ "elements, but got: " ++ show lenA
+  where
+    lenE :: Int = length expected
+    lenA :: Int = length actual
+
+assert_shape :: forall a sh. (VS.Storable a, OS.Shape sh)
+             => (a -> a -> Assertion) -- ^ The function used to make an assertion on two elements (expected, actual)
+             -> OS.Array sh a         -- ^ The expected value
+             -> OS.Array sh a         -- ^ The actual value
+             -> Assertion
+assert_shape make_assert expected actual =
+  if shapeE == shapeA then
+    assert_list make_assert (OS.toList expected) (OS.toList actual)
+  else
+    assertFailure $ "Expected shape: " ++ show shapeE ++ ", but got: " ++ show shapeA
+  where
+    shapeE = OS.shapeL expected
```

Comment:
This will be needed for untyped tensors (`OT.Array`), for which we need an instance as well. But for shaped tensors, the types system guarantees the shapes agree.

BTW, what you wrote shows me that using a MonoTraversable instance is not enough for untyped tensors, because we don't want equate tensors that have different shapes but, by chance, the same list of elements. That's a pity.

Reply:
How about now? With my new `HasShape` class?...

Reply:
Shape handling has been unified.

Reply:
`HasShape` tersely covers the case of `Matrix` and `OT.Array` and gives a unified error message for vectors as well (which means, you can remove the instance for `VS.Vector`, right?) However, checking shapes of `OS.Array` is still spurious, because `OS.shapeL` reads the list from the type and the two type are equal there.

Reply:
Even if somebody did `unsafeCoerce` and the arrays underneath have different shapes.